File: src/midi/bridge.ts
/**
 * MIDI Strummer Bridge
 *
 * Connects Strummer events to MIDI output backends.
 */

import type { MidiBackendProtocol } from './protocol.js';
import type { Strummer, StrummerEvent, StrumNoteData } from '../core/strummer.js';
import type { NoteObject } from '../models/note.js';

/**
 * Options for the MIDI Strummer Bridge
 */
export interface MidiStrummerBridgeOptions {
  /**
   * Default note duration in seconds
   */
  noteDuration?: number;

  /**
   * Whether to automatically connect to strummer events
   */
  autoConnect?: boolean;
}

/**
 * Bridge that connects Strummer events to a MIDI backend.
 *
 * Listens for strum and release events from the Strummer and
 * sends corresponding MIDI note-on/note-off messages through
 * the configured backend.
 *
 * @example
 * ```typescript
 * import { Strummer } from '../core/strummer.js';
 * import { MidiStrummerBridge, RtMidiBackend } from '../midi/index.js';
 *
 * // Create strummer and backend
 * const strummer = new Strummer();
 * const backend = new RtMidiBackend({ channel: 0 });
 * await backend.connect();
 *
 * // Create bridge - automatically connects events
 * const bridge = new MidiStrummerBridge(strummer, backend);
 *
 * // Configure note duration
 * bridge.noteDuration = 2.0;
 *
 * // Now strummer.strum() calls will send MIDI notes
 * strummer.notes = [noteObject1, noteObject2];
 * const event = strummer.strum(0.5, 0.8);
 * // MIDI notes are automatically sent!
 *
 * // Clean up
 * bridge.disconnect();
 * backend.disconnect();
 * ```
 */
export class MidiStrummerBridge {
  private _strummer: Strummer;
  private _backend: MidiBackendProtocol;
  private _noteDuration: number;
  private _isConnected = false;
  private _boundEventHandler: ((event: StrummerEvent) => void) | null = null;

  // Track currently playing notes for release
  private _activeNotes: NoteObject[] = [];

  constructor(
    strummer: Strummer,
    backend: MidiBackendProtocol,
    options: MidiStrummerBridgeOptions = {}
  ) {
    this._strummer = strummer;
    this._backend = backend;
    this._noteDuration = options.noteDuration ?? 1.5;

    if (options.autoConnect !== false) {
      this.connect();
    }
  }

  /**
   * Get the note duration
   */
  get noteDuration(): number {
    return this._noteDuration;
  }

  /**
   * Set the note duration
   */
  set noteDuration(value: number) {
    this._noteDuration = value;
  }

  /**
   * Check if the bridge is connected to strummer events
   */
  get isConnected(): boolean {
    return this._isConnected;
  }

  /**
   * Connect to strummer events
   */
  connect(): void {
    if (this._isConnected) {
      return;
    }

    this._boundEventHandler = this._handleStrummerEvent.bind(this);
    this._strummer.on('strum', this._boundEventHandler);
    this._strummer.on('release', this._boundEventHandler);
    this._isConnected = true;
  }

  /**
   * Disconnect from strummer events
   */
  disconnect(): void {
    if (!this._isConnected || !this._boundEventHandler) {
      return;
    }

    this._strummer.off('strum', this._boundEventHandler as (data: unknown) => void);
    this._strummer.off('release', this._boundEventHandler as (data: unknown) => void);
    this._boundEventHandler = null;
    this._isConnected = false;
  }

  /**
   * Release all currently playing notes
   */
  releaseAll(): void {
    if (this._activeNotes.length > 0) {
      this._backend.releaseNotes(this._activeNotes);
      this._activeNotes = [];
    }
    this._backend.releaseAll();
  }

  /**
   * Handle strummer events
   */
  private _handleStrummerEvent(event: StrummerEvent): void {
    if (!this._backend.isConnected) {
      return;
    }

    if (event.type === 'strum') {
      this._handleStrum(event.notes);
    } else if (event.type === 'release') {
      this._handleRelease();
    }
  }

  /**
   * Handle strum event - send MIDI notes
   */
  private _handleStrum(notes: StrumNoteData[]): void {
    // Track active notes
    this._activeNotes = notes.map((n) => n.note);

    // Send each note
    for (const noteData of notes) {
      this._backend.sendNote(noteData.note, noteData.velocity, this._noteDuration);
    }
  }

  /**
   * Handle release event
   */
  private _handleRelease(): void {
    // Release active notes
    if (this._activeNotes.length > 0) {
      this._backend.releaseNotes(this._activeNotes);
      this._activeNotes = [];
    }
  }

  /**
   * Manually send a note through the bridge
   */
  sendNote(note: NoteObject, velocity: number, duration?: number): void {
    if (!this._backend.isConnected) {
      return;
    }
    this._backend.sendNote(note, velocity, duration ?? this._noteDuration);
  }

  /**
   * Send pitch bend through the bridge
   */
  sendPitchBend(bendValue: number): void {
    if (!this._backend.isConnected || !this._backend.sendPitchBend) {
      return;
    }
    this._backend.sendPitchBend(bendValue);
  }
}

File: python/sketchatone/midi/bridge.py
"""
MIDI Strummer Bridge

Connects Strummer events to MIDI output backends.
"""

from typing import Optional, List, Dict, Any
from ..strummer.strummer import Strummer
from ..models.note import NoteObject
from .protocol import MidiBackendProtocol


class MidiStrummerBridge:
    """
    Bridge that connects Strummer events to a MIDI backend.
    
    Listens for strum and release events from the Strummer and
    sends corresponding MIDI note-on/note-off messages through
    the configured backend.
    
    Example:
        from sketchatone import Strummer
        from sketchatone.midi import MidiStrummerBridge, RtMidiBackend
        
        # Create strummer and backend
        strummer = Strummer()
        backend = RtMidiBackend(channel=1)
        backend.connect()
        
        # Create bridge - automatically connects events
        bridge = MidiStrummerBridge(strummer, backend)
        
        # Configure note duration
        bridge.note_duration = 2.0
        
        # Now strummer.strum() calls will send MIDI notes
        strummer.notes = [NoteObject(...), ...]
        event = strummer.strum(x=0.5, pressure=0.8)
        # MIDI notes are automatically sent!
        
        # Clean up
        bridge.disconnect()
        backend.disconnect()
    """
    
    def __init__(self, strummer: Strummer, backend: MidiBackendProtocol,
                 note_duration: float = 1.5, auto_connect: bool = True):
        """
        Initialize the bridge.
        
        Args:
            strummer: The Strummer instance to listen to
            backend: The MIDI backend to send notes through
            note_duration: Default duration for notes in seconds
            auto_connect: If True, automatically connect event listeners
        """
        self._strummer = strummer
        self._backend = backend
        self._note_duration = note_duration
        self._connected = False
        self._active_notes: List[NoteObject] = []
        
        if auto_connect:
            self.connect()
    
    @property
    def strummer(self) -> Strummer:
        """Get the connected Strummer instance."""
        return self._strummer
    
    @property
    def backend(self) -> MidiBackendProtocol:
        """Get the MIDI backend."""
        return self._backend
    
    @property
    def note_duration(self) -> float:
        """Get/set the default note duration in seconds."""
        return self._note_duration
    
    @note_duration.setter
    def note_duration(self, value: float) -> None:
        self._note_duration = max(0.1, value)
    
    @property
    def is_connected(self) -> bool:
        """Check if the bridge is connected to strummer events."""
        return self._connected
    
    def connect(self) -> None:
        """Connect event listeners to the strummer."""
        if self._connected:
            return
        
        self._strummer.on('strum', self._on_strum)
        self._strummer.on('release', self._on_release)
        self._connected = True
    
    def disconnect(self) -> None:
        """Disconnect event listeners and release any active notes."""
        if not self._connected:
            return
        
        # Release any active notes
        if self._active_notes:
            self._backend.release_notes(self._active_notes)
            self._active_notes.clear()
        
        self._strummer.off('strum', self._on_strum)
        self._strummer.off('release', self._on_release)
        self._connected = False
    
    def _on_strum(self, event: Dict[str, Any]) -> None:
        """Handle strum events from the strummer."""
        if event.get('type') != 'strum':
            return
        
        notes_data = event.get('notes', [])
        
        for note_data in notes_data:
            note = note_data.get('note')
            velocity = note_data.get('velocity', 100)
            
            if note is None:
                continue
            
            # Track active note
            self._active_notes.append(note)
            
            # Send MIDI note
            self._backend.send_note(
                note=note,
                velocity=velocity,
                duration=self._note_duration
            )
    
    def _on_release(self, event: Dict[str, Any]) -> None:
        """Handle release events from the strummer."""
        if event.get('type') != 'release':
            return
        
        # Release all active notes immediately
        if self._active_notes:
            self._backend.release_notes(self._active_notes)
            self._active_notes.clear()
    
    def process_strum(self, x: float, pressure: float) -> Optional[Dict[str, Any]]:
        """
        Convenience method to process a strum and send MIDI in one call.
        
        This is useful when you want to manually control the strum processing
        rather than relying on the event system.
        
        Args:
            x: X position (0 to 1)
            pressure: Pressure (0 to 1)
            
        Returns:
            The strum event if triggered, None otherwise
        """
        event = self._strummer.strum(x, pressure)
        
        if event:
            if event.get('type') == 'strum':
                self._on_strum(event)
            elif event.get('type') == 'release':
                self._on_release(event)
        
        return event
    
    def release_all(self) -> None:
        """Release all currently active notes."""
        if self._active_notes:
            self._backend.release_notes(self._active_notes)
            self._active_notes.clear()
